pkg/handlers/docker/mutation/customimage/inject_worker.go

```diff
@@ -0,0 +1,160 @@
+	"github.com/d2iq-labs/capi-runtime-extensions/common/pkg/capi/clustertopology/patches"
+	"github.com/d2iq-labs/capi-runtime-extensions/common/pkg/capi/clustertopology/patches/selectors"
+	"github.com/d2iq-labs/capi-runtime-extensions/common/pkg/capi/clustertopology/variables"
+	capdv1 "github.com/d2iq-labs/capi-runtime-extensions/common/pkg/external/sigs.k8s.io/cluster-api/test/infrastructure/docker/api/v1beta1"
+	"github.com/d2iq-labs/capi-runtime-extensions/pkg/handlers"
+	dockerclusterconfig "github.com/d2iq-labs/capi-runtime-extensions/pkg/handlers/docker/clusterconfig"
+	dockerworkerconfig "github.com/d2iq-labs/capi-runtime-extensions/pkg/handlers/docker/workerconfig"
+	"github.com/d2iq-labs/capi-runtime-extensions/pkg/handlers/generic/clusterconfig"
+	"github.com/d2iq-labs/capi-runtime-extensions/pkg/handlers/generic/workerconfig"
+)
+
+type customImageWorkerPatchHandler struct {
+	variableName      string
+	variableFieldPath []string
+}
+
+var (
+	_ commonhandlers.Named     = &customImageWorkerPatchHandler{}
+	_ mutation.GeneratePatches = &customImageWorkerPatchHandler{}
+	_ mutation.MetaMutator     = &customImageWorkerPatchHandler{}
+)
+
+func NewWorkerPatch() *customImageWorkerPatchHandler {
+	return newcustomImageWorkerPatchHandler(VariableName)
+}
+
+func NewWorkerInClusterConfigMetaPatch() *customImageControlPlanePatchHandler {
+	return newCustomImageControlPlanePatchHandler(
+		clusterconfig.MetaVariableName,
+		clusterconfig.MetaWorkerConfigName,
+		dockerclusterconfig.DockerVariableName,
+		VariableName,
+	)
+}
+
+func NewMetaWorkerPatch() *customImageWorkerPatchHandler {
+	return newcustomImageWorkerPatchHandler(
+		workerconfig.MetaVariableName,
+		dockerworkerconfig.DockerVariableName,
+		VariableName,
+	)
+}
+
+func newcustomImageWorkerPatchHandler(
+	variableName string,
+	variableFieldPath ...string,
+) *customImageWorkerPatchHandler {
+	return &customImageWorkerPatchHandler{
+		variableName:      variableName,
+		variableFieldPath: variableFieldPath,
+	}
+}
+
+func (h *customImageWorkerPatchHandler) Name() string {
+	return HandlerNamePatch
+}
+
+func (h *customImageWorkerPatchHandler) Mutate(
+	ctx context.Context,
+	obj runtime.Object,
+	vars map[string]apiextensionsv1.JSON,
+	holderRef runtimehooksv1.HolderReference,
+	_ client.ObjectKey,
+) error {
+	log := ctrl.LoggerFrom(ctx).WithValues(
+		"holderRef", holderRef,
+	)
+
+	customImageVar, found, err := variables.Get[v1alpha1.OCIImage](
+		vars,
+		h.variableName,
+		h.variableFieldPath...,
+	)
+	if err != nil {
+		return err
+	}
+	if !found {
+		log.V(5).
+			Info("Docker customImage variable not defined for workers, using default KinD node image")
+	}
+
+	log = log.WithValues(
+		"variableName",
+		h.variableName,
+		"variableFieldPath",
+		h.variableFieldPath,
+		"variableValue",
+		customImageVar,
+	)
+
+	return patches.Generate(
+		obj,
+		vars,
+		&holderRef,
+		selectors.InfrastructureWorkerMachineTemplates("v1beta1", "DockerMachineTemplate"),
```

Comment:
:( I don't really understand how this selector works, why is the worker patch getting called with control-plane vars (also why the unit tests are failing):
```
  topology:
    class: docker-quick-start
    controlPlane:
      metadata: {}
      replicas: 1
    variables:
      - name: clusterConfig
        value:
          addons:
            cni:
              provider: calico
            nfd: {}
          docker: {}
          controlPlane:
            docker:
              customImage: ghcr.io/mesosphere/kind-node:v1.2.3-cp
          workers:
            docker:
              customImage: ghcr.io/mesosphere/kind-node:v1.2.3-worker
    version: v1.27.5
    workers:
      machineDeployments:
        - class: default-worker
          name: md-0
          replicas: 1
        - class: default-worker
          name: md-1
          replicas: 1
          variables:
            overrides:
              - name: workerConfig
                value:
                  docker:
                    customImage: ghcr.io/mesosphere/kind-node:v1.2.3-worker-md-1
```

```
- apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
  kind: DockerMachineTemplate
  metadata:
    labels:
      topology.cluster.x-k8s.io/deployment-name: md-1
    name: docker-quick-start-md-1-infra-6fdjg
  spec:
    template:
      spec:
        customImage: ghcr.io/mesosphere/kind-node:v1.2.3-worker-md-1
        extraMounts:
        - containerPath: /var/run/docker.sock
          hostPath: /var/run/docker.sock
- apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
  kind: DockerMachineTemplate
  metadata:
    labels:
      topology.cluster.x-k8s.io/deployment-name: md-0
    name: docker-quick-start-md-0-infra-9ckdc
  spec:
    template:
      spec:
        customImage: ghcr.io/mesosphere/kind-node:v1.27.5
        extraMounts:
        - containerPath: /var/run/docker.sock
          hostPath: /var/run/docker.sock
- apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
  kind: DockerMachineTemplate
  metadata:
    name: docker-quick-start-control-plane-mq6p5
  spec:
    template:
      spec:
        customImage: ghcr.io/mesosphere/kind-node:v1.2.3-worker
        extraMounts:
        - containerPath: /var/run/docker.sock
          hostPath: /var/run/docker.sock
```